docs/developer/best-practices/security.asciidoc

```diff
@@ -1,55 +1,129 @@
+
+=== Prototype Pollution ===
+
+Prototype Pollution is an attack that is unique to JavaScript environments. Attackers can abuse JavaScript's prototype inheritance to
+"pollute" objects in the whole application, which is often used as a vector for XSS or RCE vulnerabilities. {kib} defends against this by
+hardening particularly sensitive functions (such as those exposed by `child_process`), and by requiring validation on all HTTP routes by
+default.
+
+*Best practices:*
+
+* Check for instances of `anObject[a][b] = c` where `a`, `b`, and `c` are controlled by user input. This includes code paths where the following
+logical code steps could be performed in separate files by completely different operations, or recursively using dynamic operations.
+* Validate any user input, including API url parameters, query parameters, and payloads. Preferably, use a schema which only allows specific
+keys/values. At a very minimum, implement a deny-list that prevents `__proto__` and `prototype.constructor` from being used within object
+keys.
+* When calling APIs which spawn new processes or potentially perform code generation from strings, defensively protect against Prototype
+Pollution by checking `Object.hasOwnProperty` if the arguments to the APIs originate from an Object. An example is the defunct Code app's
+https://github.com/elastic/kibana/blob/b49192626a8528af5d888545fb14cd1ce66a72e7/x-pack/legacy/plugins/code/server/lsp/workspace_command.ts#L40-L44[`spawnProcess`]
+function.
+** Common Node.js offenders: `child_process.spawn`, `child_process.exec`, `eval`, `Function('some string')`, `vm.runInContext(x)`,
+`vm.runInNewContext(x)`, `vm.runInThisContext()`
+** Common client-side offenders: `eval`, `Function('some string')`, `setTimeout('some string', num)`, `setInterval('some string', num)`
+
+See also:
+
+* https://portswigger.net/daily-swig/prototype-pollution-the-dangerous-and-underrated-vulnerability-impacting-javascript-applications[Prototype
+pollution: The dangerous and underrated vulnerability impacting JavaScript applications | portswigger.net]
+* https://github.com/HoLyVieR/prototype-pollution-nsec18/blob/master/paper/JavaScript_prototype_pollution_attack_in_NodeJS.pdf[Prototype
+pollution attack in NodeJS application | Olivier Arteau]
+
+=== Server-Side Request Forgery (SSRF) ===
+
+https://owasp.org/www-community/attacks/Server_Side_Request_Forgery[_OWASP reference for SSRF_]
+
+SSRF is a class of attacks where a vulnerable server is forced to make an unintended request, usually to an HTTP API. This is often used as
+a vector for information disclosure or injection attacks.
+
+*Best practices:*
+
+* Ensure that all outbound requests from the {kib} server use hard-coded URLs if possible.
+* If user input is used to construct a URL for an outbound request, ensure that an allow-list is used to validate the endpoints and that user input is escaped properly. Ideally
```

Comment:
__note:__ URLs are super hard to validate properly (relative, absolute, non-web protocols, schema-less URLs, URLs with embedded credentials, non-ASCII domain names and so on), so maybe it'd make sense to encourage developers to use exact match in the allow-list logic whenever it's possible and use RegEx/URL parsing only as the last resort?

Reply:
Yeah, I agree, I'll reword this!